keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -544,9 +601,10 @@ developers to test their validation rules.
 
 #### Beta
```

Comment:
I can understand if `oldSelf` support doesn't make the first alpha, but I see comparison against a previous value as extremely important for a beta.  It's one of the most powerful pieces of update validation we use on ratcheting fields.

Reply:
I'm pretty skeptical people would be successful writing ratcheting validation rules this way, and would strongly prefer adding ratcheting support in a way that 1) works automatically, 2) works for declarative openapi validation rules (e.g. https://github.com/kubernetes/kubernetes/issues/94060#issuecomment-677644185).

If we did end up providing the `oldSelf` object, I think we could only reliably provide it at the root of the object (or possibly for field paths that didn't have any list properties as ancestors)

Reply:
I honestly just haven't though enough about this one, and since we punted on this for the 1st alpha. I'll add an item to the graduation criteria so that we don't go to beta until we've resolved the topic.

Reply:
> 1) works automatically, 2) works for declarative openapi validation rules (e.g. [kubernetes/kubernetes#94060 (comment)](https://github.com/kubernetes/kubernetes/issues/94060#issuecomment-677644185)).

Working with this does not preclude also allowing free-form expression.  If we do change the validation semantics of the CEL expression to be run against old and new to diff the errors between them, I think that also needs resolution before calling a level beta.



> If we did end up providing the `oldSelf` object, I think we could only reliably provide it at the root of the object (or possibly for field paths that didn't have any list properties as ancestors)

I think having it on the root would be enough.





> I'll add an item to the graduation criteria to resolve the topic before beta.

Thanks.